mapper-pojo/src/main/java/org/hibernate/search/v6poc/entity/pojo/mapping/definition/annotation/impl/AnnotationMappingDefinitionImpl.java

```diff
@@ -49,21 +51,56 @@ public AnnotationMappingDefinition add(Set<Class<?>> annotatedTypes) {
 
 	@Override
 	public void contribute(BuildContext buildContext, TypeMetadataCollector collector) {
-		BeanResolver beanResolver = buildContext.getServiceManager().getBeanResolver();
+		DiscoveringContributor contributor = new DiscoveringContributor( mapperFactory, buildContext, collector );
 		annotatedTypes.stream()
 				.map( introspector::getTypeModel )
-				// Take super types into account
-				// Note: the order of super types (ascending or descending) does not matter here, we just pick one order
-				.flatMap( PojoRawTypeModel::getDescendingSuperTypes )
-				.distinct()
-				// TODO filter out standard Java types, e.g. Object or standard Java interfaces such as Serializable?
-				.forEach( typeModel -> {
-					String indexName = typeModel.getAnnotationByType( Indexed.class )
-							.map( Indexed::index ).orElse( null );
-
-					collector.collect( mapperFactory, typeModel, indexName,
-							new AnnotationPojoTypeNodeMetadataContributorImpl( beanResolver, typeModel ) );
-				} );
+				.forEach( contributor::contributeTypeAndSuperTypes );
 	}
 
+	/**
+	 * A metadata contributor that will call itself recursively on newly discovered types.
+	 */
+	private static class DiscoveringContributor implements PojoMappedTypeListener {
```

Comment:
The name of this class is very vague, even for an internal class.

Reply:
Changed to `LazilyDiscoveringMetadataContributor`. If it's still too vague, well... I don't know :/